scala_proto/scala_proto.bzl

```diff
@@ -594,3 +598,109 @@ def scalapb_proto_library(
         visibility = visibility,
     )
+
+def _scala_proto_gen_attrs(deps_providers):
+    return {
+        "deps": attr.label_list(mandatory = True, providers = deps_providers),
+        "blacklisted_protos" : attr.label_list(providers = [["proto"]]),
+        "flags": attr.string_list(default = []),
+        "plugin": attr.label(executable = True, cfg = "host"),
+        "_protoc": attr.label(executable = True, cfg = "host", default = "@com_google_protobuf//:protoc")
+    }
+
+_scala_proto_gen_outputs = {
+    "srcjar": "lib%{name}.srcjar",
+}
+
+def _scalapb_proto_gen_with_jvm_deps_impl(ctx):
+    jvm_deps = [p for p in ctx.attr.deps if hasattr(p, "proto") == False]
+
+    if "java_conversions" in ctx.attr.flags and len(jvm_deps) == 0:
+        fail("must have at least one jvm dependency if with_java is True (java_conversions is turned on)")
+
+    _scala_proto_gen_impl(ctx)
+
+    deps_jars = collect_jars(jvm_deps)
+
+    srcjarsattr = struct(srcjar = ctx.outputs.srcjar)
+    scalaattr = struct(
+        outputs = None,
+        compile_jars = deps_jars.compile_jars,
+        transitive_runtime_jars = deps_jars.transitive_runtime_jars,
+    )
+    java_provider = create_java_provider(scalaattr, depset())
+    return struct(
+        scala = scalaattr,
+        providers = [java_provider],
+        srcjars = srcjarsattr,
+    )
+
+_scalapb_proto_gen_with_jvm_deps = rule(
+    _scalapb_proto_gen_with_jvm_deps_impl,
+    attrs = _scala_proto_gen_attrs([["proto"], [JavaInfo]]),
+    outputs = _scala_proto_gen_outputs,
+)
+
+def _strip_root(file, roots):
+    """Strip first matching root which comes from proto_library(proto_source_root)
+        It assumes that proto_source_root are unique.
+        It should go away once generation is moved to aspects and roots can be handled for each proto_library individualy.
+    """
+    for root in roots:
+        prefix = root + "/" if file.is_source else file.root.path + "/" + root + "/"
+        if file.path.startswith(prefix):
+            return file.path.replace(prefix, "")
+    return file.short_path
+
+def _scala_proto_gen_impl(ctx):
+    protos = [p for p in ctx.attr.deps if hasattr(p, "proto")] # because scalapb_proto_library passes JavaInfo as well
+    descriptors = depset([f for dep in protos for f in dep.proto.transitive_descriptor_sets]).to_list()
+    sources = depset([f for dep in protos for f in dep.proto.transitive_sources]).to_list()
+    roots = depset([f for dep in protos for f in dep.proto.transitive_proto_path]).to_list()
+    inputs = depset([_strip_root(f, roots) for f in _retained_protos(sources, ctx.attr.blacklisted_protos)]).to_list()
+
+    srcdotjar = ctx.actions.declare_file("_" + ctx.label.name + "_src.jar")
+
+    ctx.actions.run(
+        inputs = [ctx.executable._protoc, ctx.executable.plugin] + descriptors,
+        outputs = [srcdotjar],
+        arguments = [
+            "--plugin=protoc-gen-scala=" + ctx.executable.plugin.path,
```

Comment:
This is going to prevent using workers right? So you have to spin up a new JVM for each file, right?

Reply:
Yes it would prevent using workers. I thought this is what is happening now as well. Now I think my assumptions might be wrong as `protoc-bridge` is doing some tricks to reuse jvm. I'll dig more and update.